packages/stitch/tests/eagerReturn.test.ts

```diff
@@ -0,0 +1,185 @@
+    { id: '2', productDealId: null, restaurantId: '2' },
+  ];
+
+  const postsSchema = makeExecutableSchema({
+    typeDefs: `
+      type Post {
+        id: ID!
+        productDealId: ID
+        restaurantId: ID
+      }
+
+      type Query {
+        post(id: ID!): Post
+      }
+    `,
+    resolvers: {
+      Query: {
+        post: (root, { id }) => posts.find(post => post.id === id),
+      }
+    }
+  });
+
+  const metadataSchema = makeExecutableSchema({
+    typeDefs: `
+      type ProductDeal {
+        id: ID!
+      }
+      type Restaurant {
+        id: ID!
+      }
+
+      type Post {
+        id: ID!
+        productDeal: ProductDeal
+        restaurant: Restaurant
+      }
+
+      input PostInput {
+        id: ID!
+        productDealId: ID
+        restaurantId: ID
+      }
+
+      type Query {
+        _posts(representations: [PostInput!]!): [Post]!
+      }
+    `,
+    resolvers: {
+      Query: {
+        _posts: (root, { representations }) => {
+          metadataRequests += 1;
+          return representations.map(({ id, productDealId, restaurantId }) => ({
+            id,
+            productDeal: productDealId ? { id: productDealId } : null,
+            restaurant: restaurantId ? { id: restaurantId } : null,
+          }));
+        },
+      }
+    }
+  });
+
+  const getGatewaySchema = (eagerReturn): GraphQLSchema => stitchSchemas({
+    subschemas: [
+      {
+        schema: postsSchema,
+        merge: {
+          Post: {
+            selectionSet: '{ id }',
+            fieldName: 'post',
+            args: ({ id }) => ({ id }),
+          },
+        },
+      },
+      {
+        schema: metadataSchema,
+        merge: {
+          Post: {
+            selectionSet: '{ id }',
+            fields: {
+              productDeal: { selectionSet: '{ productDealId }' },
+              restaurant: { selectionSet: '{ restaurantId }' },
+            },
+            fieldName: '_posts',
+            key: ({ id, productDealId, restaurantId }) => ({ id, productDealId, restaurantId }),
+            argsFromKeys: (representations) => ({ representations }),
+            eagerReturn,
+          },
+        },
+      },
+    ],
+    mergeTypes: true
+  });
+
+  beforeEach(() => {
+    metadataRequests = 0;
+  });
+
+  it('allows early return for empty keys', async () => {
+    const eagerReturn = (key) => null;
+
+    const result = await graphql(getGatewaySchema(eagerReturn), `
+      query {
+        post(id: 1) {
+          id
+          productDeal {
+            id
+          }
+          restaurant {
+            id
+          }
+        }
+      }
+    `);
+
+    expect(metadataRequests).toEqual(0);
+    expect(result.data.post).toEqual({
+      id: '1',
+      productDeal: null,
+      restaurant: null,
+    });
+  });
+
+  it('allows early return with custom result', async () => {
+    const eagerReturn = (key) => ({ id: key.id, restaurant: { id: '55' } });
```

Comment:
@yaacovCR – in this case, this eagerly returned object would be the literal return from `batchDelegateToSchema`... is that valid for subsequent delegations? I see delegated object results normally include symbol keys with some implementation references in them. Is this kind of direct result without going through `delegateToSchema` possible/worth supporting?

Reply:
I think we would have to call handleResult (or resolveExternalValue in v7) on the result in order to properly process the value into the correct format.

What is the use case for returning a custom result? Why program on the gateway if the fields are available from that subschema?

Reply:
The usecase would mainly be non-nullable arrays. If you have “restaurantIds: []” on the representation then you know the “restaurants” array is going to come back empty. However, if the field is “restaurants: [Restaurant]!”, then the result must include the empty array. This handler could be smart enough to translate blank key structures into blank result structures if it has to fulfill not-null fields.

That said, we could also just write off that usecase and say that not-null fields require a service request. If we did that, then “eagerReturn” would be simplified down to just a “keyIsEmpty” check that primes null for empty keys.